github/Branch.py

```diff
@@ -161,155 +169,106 @@ def edit_protection(
-            required_conversation_resolution is github.GithubObject.NotSet
-            or isinstance(required_linear_history, bool)
         ), required_conversation_resolution
-        assert lock_branch is github.GithubObject.NotSet or isinstance(
-            lock_branch, bool
-        ), lock_branch
-        assert allow_fork_syncing is github.GithubObject.NotSet or isinstance(
-            allow_fork_syncing, bool
-        ), allow_fork_syncing
-        assert (
-            users_bypass_pull_request_allowances is github.GithubObject.NotSet
-            or all(
-                isinstance(element, str)
-                for element in users_bypass_pull_request_allowances
-            )
-        ), (users_bypass_pull_request_allowances)
-        assert (
-            teams_bypass_pull_request_allowances is github.GithubObject.NotSet
-            or all(
-                isinstance(element, str)
-                for element in teams_bypass_pull_request_allowances
-            )
-        ), (teams_bypass_pull_request_allowances)
-        assert apps_bypass_pull_request_allowances is github.GithubObject.NotSet or all(
-            isinstance(element, str) for element in apps_bypass_pull_request_allowances
+        assert is_optional(lock_branch, bool), lock_branch
+        assert is_optional(allow_fork_syncing, bool), allow_fork_syncing
+        assert is_optional_list_of_type(
+            users_bypass_pull_request_allowances, str
+        ), users_bypass_pull_request_allowances
+        assert is_optional_list_of_type(
+            teams_bypass_pull_request_allowances, str
+        ), teams_bypass_pull_request_allowances
+        assert is_optional_list_of_type(
+            apps_bypass_pull_request_allowances, str
         ), apps_bypass_pull_request_allowances
 
-        post_parameters = {}
-        if (
-            strict is not github.GithubObject.NotSet
-            or contexts is not github.GithubObject.NotSet
-        ):
-            if strict is github.GithubObject.NotSet:
+        post_parameters: dict[str, Any] = {}
+
+        required_status_checks = {}
+        if is_defined(strict) or is_defined(contexts):
+            if is_undefined(strict):
                 strict = False
-            if contexts is github.GithubObject.NotSet:
+            if is_undefined(contexts):
                 contexts = []
-            post_parameters["required_status_checks"] = {
+            required_status_checks = {
                 "strict": strict,
                 "contexts": contexts,
             }
-        else:
-            post_parameters["required_status_checks"] = None
 
-        if enforce_admins is not github.GithubObject.NotSet:
+        post_parameters["required_status_checks"] = required_status_checks or None
+
+        if is_defined(enforce_admins):
             post_parameters["enforce_admins"] = enforce_admins
         else:
             post_parameters["enforce_admins"] = None
 
+        required_pr_reviews: dict[str, Any] = {}
```

Comment:
minor refactor here, avoid `... or ... or ... or ... or ... or ...`

Reply:
This touches a lot of logic, we'll have to carefully review this.

Reply:
I can revert them to previous logic, only change `... is NotSet`

Reply:
Lets not rewrite the logic, stick to merging typing into source, which only includes the `NotSet` checking. The rest of the improvements should be done in a separate PR (e.g. `None` defaults).